File: src/training/pipeline/zig/src/schema_extractor.zig
const std = @import("std");
const csv_parser = @import("csv_parser.zig");
const schema_registry = @import("schema_registry.zig");

pub fn extractFromStagingCsv(
    allocator: std.mem.Allocator,
    csv_data: []const u8,
    registry: *schema_registry.SchemaRegistry,
) !void {
    var parser = csv_parser.CsvParser.init(allocator, csv_data);

    // Skip header rows (first 3 rows are metadata headers)
    var headers_skipped: u32 = 0;
    while (headers_skipped < 3) {
        if (try parser.nextRow()) |row_val| {
            var row = row_val;
            row.deinit();
        } else return;
        headers_skipped += 1;
    }

    // Track tables we've seen to avoid duplicates
    var seen_tables = std.StringHashMap(usize).init(allocator);
    defer {
        var it = seen_tables.keyIterator();
        while (it.next()) |key| {
            allocator.free(key.*);
        }
        seen_tables.deinit();
    }

    while (try parser.nextRow()) |row_val| {
        var row = row_val;
        defer row.deinit();

        if (row.fields.len < 10) continue;

        const schema_name = row.fields[5]; // BTP Staging Schema Name
        const table_name = row.fields[6]; // BTP Table Name
        const field_name = row.fields[7]; // BTP Field Name

        if (table_name.len == 0 or field_name.len == 0) continue;

        // Get or create table index
        if (seen_tables.get(table_name) == null) {
            const idx = registry.tables.items.len;
            try registry.addTable(.{
                .name = try allocator.dupe(u8, table_name),
                .schema_name = try allocator.dupe(u8, schema_name),
                .domain = domainFromUseCase(if (row.fields.len > 1) row.fields[1] else ""),
                .columns = &.{},
                .hierarchy_levels = &.{},
                .row_count = 0,
                .description = try allocator.dupe(u8, ""),
            });
            try seen_tables.put(try allocator.dupe(u8, table_name), idx);
        }
    }
}

fn domainFromUseCase(use_case: []const u8) schema_registry.Domain {
    if (std.mem.indexOf(u8, use_case, "TREASURY") != null or
        std.mem.indexOf(u8, use_case, "CAPITAL") != null)
    {
        return .TREASURY;
    }
    if (std.mem.indexOf(u8, use_case, "ESG") != null) {
        return .ESG;
    }
    return .PERFORMANCE;
}

test "extract tables from staging csv" {
    const allocator = std.testing.allocator;
    const csv_data =
        \\header1
        \\header2
        \\header3
        \\,TREASURY_CAPITAL,BCRS,"TABLE1",AS_OF_DATE,STG_BCRS,BSI_REM_FACT,AS_OF_DATE,Date field,TIMESTAMP,,,
        \\,TREASURY_CAPITAL,BCRS,"TABLE1",STATUS,STG_BCRS,BSI_REM_FACT,STATUS,Status field,NVARCHAR,,,
        \\,TREASURY_CAPITAL,BCRS,"TABLE2",COUNTRY,STG_BCRS,BSI_REM_DIM_COUNTRY,COUNTRY,Country name,NVARCHAR,,,
    ;

    var registry = schema_registry.SchemaRegistry.init(allocator);
    defer {
        // Free allocated table names and schema names
        for (registry.tables.items) |t| {
            allocator.free(t.name);
            allocator.free(t.schema_name);
            allocator.free(t.description);
        }
        registry.deinit();
    }

    try extractFromStagingCsv(allocator, csv_data, &registry);

    // Should have extracted 2 tables
    try std.testing.expectEqual(@as(usize, 2), registry.tables.items.len);
    try std.testing.expectEqualStrings("BSI_REM_FACT", registry.tables.items[0].name);
    try std.testing.expectEqualStrings("STG_BCRS", registry.tables.items[0].schema_name);
}

File: src/training/pipeline/schema_extractor.py
# =============================================================================
# schema_extractor.py — Extract table schemas from staging CSV files
# =============================================================================
from __future__ import annotations

from pathlib import Path

from .csv_parser import parse_csv_file, parse_csv_string
from .schema_registry import Column, Domain, SchemaRegistry, TableSchema


def _domain_from_use_case(use_case: str) -> Domain:
    """Infer domain from the use-case column value."""
    upper = use_case.upper()
    if "TREASURY" in upper or "CAPITAL" in upper:
        return Domain.TREASURY
    if "ESG" in upper:
        return Domain.ESG
    return Domain.PERFORMANCE


def extract_from_staging_csv(csv_path: str | Path, registry: SchemaRegistry) -> None:
    """Extract table schemas from a staging CSV file into the registry.

    Expects the first 3 rows to be metadata headers (skipped).
    Data rows are expected to have at least 10 columns with fields:
      [5] = BTP Staging Schema Name
      [6] = BTP Table Name
      [7] = BTP Field Name
      [8] = Field Description
      [9] = Data Type
    """
    rows = parse_csv_file(csv_path)
    _extract_rows(rows, registry)


def extract_from_staging_csv_string(csv_data: str, registry: SchemaRegistry) -> None:
    """Same as extract_from_staging_csv but from an in-memory string."""
    rows = parse_csv_string(csv_data)
    _extract_rows(rows, registry)


def _extract_rows(rows: list, registry: SchemaRegistry) -> None:
    """Internal: process parsed CSV rows into the schema registry."""
    # Skip 3 metadata header rows
    data_rows = rows[3:] if len(rows) > 3 else []

    for row in data_rows:
        if len(row.fields) < 10:
            continue

        use_case = row.fields[1] if len(row.fields) > 1 else ""
        schema_name = row.fields[5]
        table_name = row.fields[6]
        field_name = row.fields[7]
        field_desc = row.fields[8] if len(row.fields) > 8 else ""
        data_type = row.fields[9] if len(row.fields) > 9 else "NVARCHAR"

        if not table_name or not field_name:
            continue

        # Add table if not already registered
        registry.add_table(
            TableSchema(
                name=table_name,
                schema_name=schema_name,
                domain=_domain_from_use_case(use_case),
            )
        )

        # Add column to the table
        registry.add_column(
            table_name,
            Column(
                name=field_name,
                data_type=data_type,
                description=field_desc,
            ),
        )
